R/make-seurat.R

```diff
@@ -0,0 +1,167 @@
+#' @param dedup_method Method to handle duplicated gene symbols. If `unique`,
+#'   the gene symbols will be made unique following standard Seurat procedures.
+#'   If `sum`, the expression values for each gene symbols will be summed.
+#' @param seurat_assay_version Whether to create Seurat `v3` or `v5` assay
+#'   objects. Default is `v3`.
+#'
+#' @return a Seurat object
+#' @export
+#'
+#' @examples
+#' \dontrun{
+#' # convert an ScPCA SCE to Seurat renaming with gene symbols
+#' seurat_obj <- sce_to_seurat(sce)
+#'
+#' # convert SCE to Seurat keeping the current names (usually Ensembl)
+#' seurat_obj <- sce_to_seurat(sce, use_symbols = FALSE)
+#'
+#' # convert SCE to Seurat, merging duplicated gene names
+#' seurat_obj <- sce_to_seurat(sce, dedup_method = "sum")
+#'
+#' # convert SCE to Seurat with v5 assay objects
+#' seurat_obj <- sce_to_seurat(sce, seurat_assay_version = "v5")
+#' }
+#'
+sce_to_seurat <- function(
+    sce,
+    use_symbols = TRUE,
+    reference = c("sce", "scpca", "10x2020", "10x2024"),
+    dedup_method = c("unique", "sum"),
+    seurat_assay_version = c("v3", "v5")) {
+  reference <- match.arg(reference)
+  dedup_method <- match.arg(dedup_method)
+  seurat_assay_version <- match.arg(seurat_assay_version)
+
+  stopifnot(
+    "Package `Seurat` must be installed for SCE to Seurat conversion." =
+      requireNamespace("Seurat", quietly = TRUE),
+    "Package `SeuratObject` must be installed for SCE to Seurat conversion." =
+      requireNamespace("SeuratObject", quietly = TRUE),
+    "`sce` must be a SingleCellExperiment object." = is(sce, "SingleCellExperiment"),
+    "`sce` must contain both `gene_ids` and `gene_symbol` columns in the row data if it is being used as a reference." =
+      reference != "sce" || all(c("gene_ids", "gene_symbol") %in% names(rowData(sce)))
+  )
+
+  if (use_symbols) {
+    sce <- suppressMessages(sce_to_symbols(
+      sce,
+      reference = reference,
+      unique = (dedup_method == "unique"),
+      seurat_compatible = TRUE
+    ))
+  }
+
+  if (dedup_method == "sum") {
+    sce <- sum_duplicate_genes(sce)
+  }
+
+  mainExpName(sce) <- "RNA"
+
+  # convert reducedDims to Seurat standard naming
+  reducedDimNames(sce) <- tolower(reducedDimNames(sce))
+  reducedDims(sce) <- reducedDims(sce) |>
+    purrr::map(\(mat) {
+      # insert underscore between letters and numbers in axis labels
+      rd_names <- gsub("([A-Za-z]+)([0-9]+)", "\\1_\\2", colnames(mat)) |>
+        tolower()
+      colnames(mat) <- rd_names
+      if (!is.null(attr(mat, "rotation"))) {
+        colnames(attr(mat, "rotation")) <- rd_names
+      }
+      return(mat)
+    })
+
+  # pull out any altExps to handle manually
+  alt_exps <- altExps(sce)
+  altExps(sce) <- NULL
+
+  # Let Seurat do initial conversion to capture most things
+  if ("logcounts" %in% assayNames(sce)) {
+    data_name <- "logcounts"
+  } else {
+    data_name <- NULL
+  }
+  sobj <- Seurat::as.Seurat(sce, data = data_name)
```

Comment:
Tests clearly pass, but I'm unable to run `sce_to_seurat()` standalone and the error is showing up at this line, both with a processed and filtered SCE. Attaching Seurat(Object) didn't help. Are you able to run this outside of the test suite?
```
Error in UseMethod(generic = "DefaultAssay<-", object = object) : 
  no applicable method for 'DefaultAssay<-' applied to an object of class "call"
```

```
> sessionInfo()
R version 4.4.0 (2024-04-24)
Platform: aarch64-apple-darwin20
Running under: macOS 15.1

Matrix products: default
BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/New_York
tzcode source: internal

attached base packages:
[1] stats4    stats     graphics  grDevices datasets  utils     methods   base     

other attached packages:
 [1] rOpenScPCA_0.1.0            SingleCellExperiment_1.26.0 SummarizedExperiment_1.34.0
 [4] Biobase_2.64.0              GenomicRanges_1.56.1        GenomeInfoDb_1.40.1        
 [7] IRanges_2.38.1              S4Vectors_0.42.1            BiocGenerics_0.50.0        
[10] MatrixGenerics_1.16.0       matrixStats_1.4.1           Seurat_5.1.0               
[13] SeuratObject_5.0.2          sp_2.1-4                    testthat_3.2.2             

loaded via a namespace (and not attached):
  [1] RcppAnnoy_0.0.22        splines_4.4.0           later_1.3.2            
  [4] tibble_3.2.1            polyclip_1.10-7         fastDummies_1.7.4      
  [7] lifecycle_1.0.4         rprojroot_2.0.4         globals_0.16.3         
 [10] lattice_0.22-6          MASS_7.3-60.2           magrittr_2.0.3         
 [13] plotly_4.10.4           yaml_2.3.10             remotes_2.5.0          
 [16] httpuv_1.6.15           sctransform_0.4.1       spam_2.10-0            
 [19] sessioninfo_1.2.2       pkgbuild_1.4.4          spatstat.sparse_3.1-0  
 [22] reticulate_1.39.0       cowplot_1.1.3           pbapply_1.7-2          
 [25] RColorBrewer_1.1-3      abind_1.4-5             pkgload_1.4.0          
 [28] zlibbioc_1.50.0         Rtsne_0.17              purrr_1.0.2            
 [31] GenomeInfoDbData_1.2.12 ggrepel_0.9.6           irlba_2.3.5.1          
 [34] listenv_0.9.1           spatstat.utils_3.1-0    goftest_1.2-3          
 [37] RSpectra_0.16-2         spatstat.random_3.3-1   fitdistrplus_1.2-1     
 [40] parallelly_1.38.0       leiden_0.4.3.1          codetools_0.2-20       
 [43] DelayedArray_0.30.1     tidyselect_1.2.1        UCSC.utils_1.0.0       
 [46] pdfCluster_1.0-4        spatstat.explore_3.3-2  jsonlite_1.8.8         
 [49] BiocNeighbors_1.22.0    ellipsis_0.3.2          progressr_0.14.0       
 [52] ggridges_0.5.6          survival_3.7-0          tools_4.4.0            
 [55] ica_1.0-3               Rcpp_1.0.13             glue_1.7.0             
 [58] gridExtra_2.3           SparseArray_1.4.8       usethis_3.0.0          
 [61] dplyr_1.1.4             withr_3.0.2             BiocManager_1.30.25    
 [64] fastmap_1.2.0           bluster_1.14.0          fansi_1.0.6            
 [67] digest_0.6.37           R6_2.5.1                mime_0.12              
 [70] colorspace_2.1-1        scattermore_1.2         tensor_1.5             
 [73] spatstat.data_3.1-2     utf8_1.2.4              tidyr_1.3.1            
 [76] generics_0.1.3          renv_1.0.11             data.table_1.16.0      
 [79] httr_1.4.7              htmlwidgets_1.6.4       S4Arrays_1.4.1         
 [82] uwot_0.2.2              pkgconfig_2.0.3         gtable_0.3.5           
 [85] lmtest_0.9-40           XVector_0.44.0          brio_1.1.5             
 [88] htmltools_0.5.8.1       profvis_0.4.0           dotCall64_1.1-1        
 [91] scales_1.3.0            png_0.1-8               spatstat.univar_3.0-1  
 [94] geometry_0.5.0          rstudioapi_0.17.1       reshape2_1.4.4         
 [97] nlme_3.1-164            magic_1.6-1             cachem_1.1.0           
[100] zoo_1.8-12              stringr_1.5.1           KernSmooth_2.23-24     
[103] parallel_4.4.0          miniUI_0.1.1.1          desc_1.4.3             
[106] pillar_1.9.0            grid_4.4.0              vctrs_0.6.5            
[109] RANN_2.6.2              urlchecker_1.0.1        promises_1.3.0         
[112] xtable_1.8-4            cluster_2.1.6           waldo_0.6.1            
[115] cli_3.6.3               compiler_4.4.0          rlang_1.1.4            
[118] crayon_1.5.3            future.apply_1.11.2     plyr_1.8.9             
[121] fs_1.6.4                stringi_1.8.4           deldir_2.0-4           
[124] viridisLite_0.4.2       BiocParallel_1.38.0     munsell_0.5.1          
[127] lazyeval_0.2.2          devtools_2.4.5          spatstat.geom_3.3-2    
[130] Matrix_1.7-0            RcppHNSW_0.6.0          patchwork_1.2.0        
[133] future_1.34.0           ggplot2_3.5.1           shiny_1.9.1            
[136] ROCR_1.0-11             igraph_2.0.3            memoise_2.0.1 
```

Reply:
https://github.com/AlexsLemonade/rOpenScPCA/pull/15/commits/914e500ffa788a4ac9307833d9a30af08c2b047c seems to have done it

Reply:
This is probably related to [an `renv` bug in the latest version](https://github.com/rstudio/renv/issues/2017); do you have a variable named `object` in your environment? (Seurat must be doing something bad here too, because it shouldn't read a global variable wherever it is doing that, but I can't work around all the Seurat bugs...) It should not affect running the function if you load the package from an external source. I just pushed a temporary fix until `renv` update comes.

Reply:
> do you have a variable named object in your environment? 

No definitely not. 

...But I just went back to that commit hash and re-ran my exact code, and uh.... magic? Clearly, black magic. But black magic from devtools or testthat. 

```
Restarting R session...

- Project '~/ALSF/open-scpca/rOpenScPCA' loaded. [renv 1.0.11]
> devtools::load_all(".")
ℹ Loading rOpenScPCA
> sce <- readRDS("tests/testthat/data/scpca_sce.rds")
> s <- sce_to_seurat(sce)
Error in UseMethod(generic = "DefaultAssay<-", object = object) : 
  no applicable method for 'DefaultAssay<-' applied to an object of class "call"
In addition: Warning message:
In ensembl_to_symbol(ensembl_ids, sce = sce, unique = unique, seurat_compatible = seurat_compatible) :
  Replacing underscores ('_') with dashes ('-') in gene symbols for Seurat compatibility.
> 
> object
test_check("rOpenScPCA")
```

I restarted R again, and...

```
Restarting R session...

- Project '~/ALSF/open-scpca/rOpenScPCA' loaded. [renv 1.0.11]
> object
test_check("rOpenScPCA")
```


I have made a short film.
https://github.com/user-attachments/assets/9b96416a-83f2-4091-b8d9-7aab22fa5d05

Reply:
Again, this was at the previous commit; this behavior no longer happens after https://github.com/AlexsLemonade/rOpenScPCA/commit/914e500ffa788a4ac9307833d9a30af08c2b047c